File: src/data_structures/tree/TreeNode.kt
package data_structures.tree

import data_structures.queues.Queues

class TreeNode<T>(val value: T) {
    private val children: MutableList<TreeNode<T>> = mutableListOf()

    fun add(child: TreeNode<T>) = children.add(child)

    fun forEachDepthFirst(visit: Visitor<T>) {
        visit(this)
        children.forEach {
            it.forEachDepthFirst(visit)
        }
    }

    fun forEachLevelOrder(visit: Visitor<T>) {
        visit(this)
        val queue = Queues<TreeNode<T>>()
        children.forEach {
            queue.enqueue(it)
        }

        var node = queue.dequeue()
        while (node != null) {
            visit(node)
            node.children.forEach {
                queue.enqueue(it)
            }
            node = queue.dequeue()
        }
    }

    fun search(value: T): TreeNode<T>? {
        var result: TreeNode<T>? = null
        forEachLevelOrder {
            if (it.value == value)
                result = it
        }
        return result
    }


}
typealias Visitor<T> = (TreeNode<T>) -> Unit

File: data_structures/tree/tree_node.py
from queue import Queue


class TreeNode:
    def __init__(self, value):
        self.children = []
        self.value = value

    def add(self, node):
        if node is not TreeNode:
            self.children.append(node)

    # Di chuyển từ trên xuống dưới và từ trái qua phải.
    # Thăm node từ gốc tới leaf trái -> phải
    def for_each_depth_first(self, visit):
        visit(self)
        for node in self.children:
            node.for_each_depth_first(visit)

    # Di chuyển theo từng tầng một
    def for_each_level_order(self, visit):
        visit(self)
        queue = Queue()
        for node in self.children:
            queue.put_nowait(node)

        node = queue.get_nowait()
        while node is not None:
            visit(node)
            for child in node.children:
                queue.put_nowait(child)
            if queue.empty():
                node = None
            else:
                node = queue.get_nowait()

    def search(self, value):
        if value == self.value:
            return self
        queue = Queue()
        for node in self.children:
            queue.put_nowait(node)

        node = queue.get_nowait()
        while node is not None:
            if node.value == value:
                return node
            for child in node.children:
                queue.put_nowait(child)
            if queue.empty():
                node = None
            else:
                node = queue.get_nowait()

        return None
